<a href="https://colab.research.google.com/github/GretRincon/PrimerProyectoFisica/blob/master/SeriePandasGret.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Serie Pandas
Esta es la segunda serie, en esta quiero que trabajemos con datasets que se obtendrán a partir del github del curso que es: https://github.com/adrishg/PythonML-PrebecariosProteco 

(No les subiré un punto por seguirme porque no estoy tan desesperada por popularidad pero deberían considerarlo)

In [0]:
import pandas as pd

1) Considerando el siguiente set de datos:
           
           a) Crea un dataframe a partir de datos
           b) Despliega los primeros tres datos de dicho dataframe
           c) Despliga los últimos dos datos de dicho dataframe
           d) Una descripción de los datos (count, mean, std, ...) 

In [0]:
datos = {
    'fecha':pd.date_range('20200430', periods=6),
    'calidad':pd.Categorical(['buena', 'muy buena', 'mala', 'buena', 'muy mala', 'buena'],
                                 categories = ['muy mala', 'mala', 'regular', 'buena', 'muy buena'],
                                 ordered = True),
    'asistentes':[10, 25, 12, 3, 23, 43],
    'recomendaciones':['hablar más lento', 'me aburri', 'Derek hacia cosas raras', 'excelente', 'no se entendio nada',
               '¿Y el recreo?']
}

In [10]:
import pandas as pd
dataFrameDatos=pd.DataFrame(datos) #Crear el dataframe (tabla de datos)
print("Esta es la tabla de los datos recolectados")
dataFrameDatos #Mandar a llamar el dataframe


Esta es la tabla de los datos recolectados


fecha    calidad  asistentes          recomendaciones
0 2020-04-30      buena          10         hablar más lento
1 2020-05-01  muy buena          25                me aburri
2 2020-05-02       mala          12  Derek hacia cosas raras
3 2020-05-03      buena           3                excelente
4 2020-05-04   muy mala          23      no se entendio nada
5 2020-05-05      buena          43            ¿Y el recreo?

In [6]:
print("Los primeros 3 datos son:")
dataFrameDatos.head(3) #Imprime los primeros n datos


Los primeros 3 datos son:
Los últimos 2 datos son:


asistentes
count    6.000000
mean    19.333333
std     14.236104
min      3.000000
25%     10.500000
50%     17.500000
75%     24.500000
max     43.000000

In [8]:
print("Los últimos 2 datos son:")
dataFrameDatos.tail(2) #Imprime los ultimos n datos


Los últimos 2 datos son:


fecha   calidad  asistentes      recomendaciones
4 2020-05-04  muy mala          23  no se entendio nada
5 2020-05-05     buena          43        ¿Y el recreo?

In [9]:
dataFrameDatos.describe() #Te da count, mean, std, min, 25%,50%,75%, max
#Count, numero de datos (sin contar los vacios)
#std desviacion estandar
#min, max, Son los valores numericos minimos y maximos dentro del dataframe
#Puedes variar los parametros dentro del .describe()

asistentes
count    6.000000
mean    19.333333
std     14.236104
min      3.000000
25%     10.500000
50%     17.500000
75%     24.500000
max     43.000000

2) Crea una función que pida número de columnas, su nombre y número de datos a ingresar de manera que el usuario pueda crear un dataframe con los datos de su preferencia.

In [15]:
import pandas as pd 
def creaDataFrames():
    Construcciondatos={} #Diccionario, importante después
    #Cuantas columnas y filas quieres
    Columnas=int(input("Ingresa el número de columnas de tu dataframe "))
    Filas=int(input("Ingresa el número de filas "))
    #Titulo a cada columna, esto implica crear listas y combinar luego en dataframes
    for i in range(Columnas):
      print("Ahora estamos en la columna ",i+1) 
      TituloColumna=input("Ingresa el título de tu columna ")
      columna=[] #Lista definida
      for j in range(Filas):
          print("Fila ",j+1) 
          dato= input("Ingresa tu dato ")
          columna.append(dato) #Agregar dato a columna por cada fila
      #Ahora que hemos terminado la columna, la agregamos a un diccionario¡ Continuamos 
      Construcciondatos[TituloColumna]=columna #Titulo columna es la clave
    dataFrameNuevo=pd.DataFrame(Construcciondatos) #Ahora si, el diccionario es un dataframe
    print(dataFrameNuevo) #Voila

print("Crea tu dataframe ") #Instrucción ¿Que hace el programa?
creaDataFrames() #Llama y comienza la funcion
       


Crea tu dataframe 
Ingresa el número de columnas de tu dataframe 3
Ingresa el número de filas 3
Ahora estamos en la columna  1
Ingresa el título de tu columna Materia
Fila  1
Ingresa tu dato Cuántica
Fila  2
Ingresa tu dato Electro
Fila  3
Ingresa tu dato Optica
Ahora estamos en la columna  2
Ingresa el título de tu columna Calificación
Fila  1
Ingresa tu dato 8
Fila  2
Ingresa tu dato 10
Fila  3
Ingresa tu dato 10
Ahora estamos en la columna  3
Ingresa el título de tu columna Profesor
Fila  1
Ingresa tu dato Bueno
Fila  2
Ingresa tu dato Barco
Fila  3
Ingresa tu dato Malo
    Materia Calificación Profesor
0  Cuántica            8    Bueno
1   Electro           10    Barco
2    Optica           10     Malo


3) A partir del siguiente DataSet realiza:

In [14]:
#Liga del github: https://raw.githubusercontent.com/adrishg/PythonML-PrebecariosProteco/master/DataSets/startupsUSA.csv
#Podrán agregar archivos de github y leerlos usando el url raw
data = pd.read_csv('https://raw.githubusercontent.com/adrishg/PythonML-PrebecariosProteco/master/DataSets/startupsUSA.csv')
data.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

    a) Crea una función que reciba un lugar y despligue en el número de empresas que residen en dicho lugar

In [18]:
def EmpresasenLugar(lugar): #Función, dame un lugar, te despliego las empresas en él
    Elegidos = data['State'] == lugar #datos donde la columna state, coincida con lugar  
    DatosEEL = data[Elegidos] #Un 2do dataframe solo con el lugar especifico
    #DatosEEL (Empresas en "lugar")
    print("El numero de empresas en",lugar,"es",len(DatosEEL)) #Cuantas en lugar
    print(DatosEEL)

EmpresasenLugar('New York') #Aqui estamos dando por defecto New York
#Podriamos desplegar un menu de los lugares New York, California,Florida


El numero de empresas en New York es 17
    R&D Spend  Administration  Marketing Spend     State     Profit
0   165349.20       136897.80        471784.10  New York  192261.83
3   144372.41       118671.85        383199.62  New York  182901.99
5   131876.90        99814.71        362861.36  New York  156991.12
8   120542.52       148718.95        311613.29  New York  152211.77
15  114523.61       122616.84        261776.23  New York  129917.04
17   94657.16       145077.58        282574.31  New York  125370.37
19   86419.70       153514.11             0.00  New York  122776.86
21   78389.47       153773.43        299737.29  New York  111313.02
24   77044.01        99281.34        140574.81  New York  108552.04
27   72107.60       127864.55        353183.81  New York  105008.31
29   65605.48       153032.06        107138.38  New York  101004.64
31   61136.38       152701.92         88218.23  New York   97483.56
35   46014.02        85047.44        205517.64  New York   96479.51
38   202

    b) Un data set cuyo orden este dado alfabeticamente por los lugares donde están las empresas

In [21]:
LugarAlfab = data.sort_values('State',ascending=True) #sort=ordenar en State de A a Z
print(LugarAlfab)

    R&D Spend  Administration  Marketing Spend       State     Profit
49       0.00       116983.80         45173.06  California   14681.40
42   23640.93        96189.63        148001.11  California   71498.49
40   28754.33       118546.05        172795.67  California   78239.91
34   46426.07       157693.92        210797.67  California   96712.80
16   78013.11       121597.55        264346.06  California  126992.93
44   22177.74       154806.14         28334.72  California   65200.33
25   64664.71       139553.16        137962.62  California  107404.34
13   91992.39       135495.07        252664.93  California  134307.35
11  100671.96        91790.61        249744.55  California  144259.40
20   76253.86       113867.30        298664.47  California  118474.03
39   38558.51        82982.09        174999.30  California   81005.76
6   134615.46       147198.87        127716.82  California  156122.51
37   44069.95        51283.14        197029.42  California   89949.14
32   63408.86       

    c) Una función donde ingreses el index númerico y despliegue dichos datos de manera legible:
    
        Ejemplo: [In]: 0  [Out]: La empresa que reside en New York tiene un gato de Marketing de 471784.10 dólares mientras que sus ganancias son 192261.83 dólares...

In [15]:
def Indice(i):
    resgistroi = data.iloc[i]
    print("La empresa que reside en", resgistroi['State'],"tiene un gasto de marketing de ", resgistroi['Marketing Spend'],"dólares mientras que sus ganancias son ", resgistroi['Profit'],"dólares"  )
Indice(3)

La empresa que reside en New York tiene un gasto de marketing de  383199.62 dólares mientras que sus ganancias son  182901.99 dólares


4) Deberán cargar los datos de iris setosa de dos maneras:

    a) Deberán dirigirse al repositorio del curso: https://github.com/adrishg/PythonML-PrebecariosProteco
    De ahí a la carpeta de DataSets, ahí encontrarán el archivo iris.csv deben darle click para previsualizar el archivo. Ahí encontrarán un botón que dice Raw denle click y copien la liga, ese es el link raw y podrán exportarlo como lo hice en la parte de arriba con los datos de las empresas estadounidenses.
    Desplieguen los datos con un describe para que revisen que se exporto bien.
    
    NOTA: También pueden aprovechar y seguirme en github, saludos.
    
![screenshot iris csv](https://drive.google.com/uc?id=1ytlucQdsbdnWsJmTG61eR7WQ_PBFNmCQ)
    

In [21]:
datairis = pd.read_csv('https://raw.githubusercontent.com/adrishg/PythonML-PrebecariosProteco/master/DataSets/iris.csv')
datairis.describe()

Id  sepal_length  sepal_width  petal_length  petal_width
count  150.000000    150.000000   150.000000    150.000000   150.000000
mean    75.500000      5.843333     3.054000      3.758667     1.198667
std     43.445368      0.828066     0.433594      1.764420     0.763161
min      1.000000      4.300000     2.000000      1.000000     0.100000
25%     38.250000      5.100000     2.800000      1.600000     0.300000
50%     75.500000      5.800000     3.000000      4.350000     1.300000
75%    112.750000      6.400000     3.300000      5.100000     1.800000
max    150.000000      7.900000     4.400000      6.900000     2.500000

    b) El dataset de Iris es muy conocido en el área ya que está precargado en sklearn.datasets. Pueden obtenerlo a partir de este, tendrán que investigarlo y exportar los datos.

In [19]:
from sklearn.datasets import load_iris #sklearn incluye diversas listas de datos como boston, breastcancer, diabetes, digits,iris y linnerud
iris = load_iris() #Ingresamos el set de datos iris, con ese mismo nombre
irisDataFrame = pd.DataFrame(iris.data, columns=iris.feature_names) #Lo convertimos a dataframe, columns lo usamos para recuperar el nombre de las columnas
irisDataFrame.describe() 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

    c) Por último les recomiendo leer acerca de este dataset, lo usaremos bastante en clase.

5) En esta opción eres libre de poner cualquier función que use elementos y funciones de la librería pandas. 
    El requerimiento es que uses iloc, puedes usar alguno de los datasets cargados en esta libreta de jupyter.


In [40]:
def llamari (frame,i): #Llamara el dato con indice i de cualquier dataframe
  print(frame.iloc[i]) 
irislength = datairis.sort_values('petal_length', ascending=True)
irislength.head()
print("Tu base de datos iris esta ordenada por la longitud de petalo ")
i=(int(input("¿Que dato quieres desplegar? Escribe el número de dato ")))
llamari(irislength,i)


Tu base de datos iris esta ordenada por la longitud de petalo 
¿Que dato quieres desplegar? Escribe el número de dato 5
Id                       41
sepal_length              5
sepal_width             3.5
petal_length            1.3
petal_width             0.3
species         Iris-setosa
Name: 40, dtype: object
